tensorflow/lite/micro/examples/hello_world/quantization/ptq.py

```diff
@@ -0,0 +1,163 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+"""hellow_world model training for sinwave recognition
+
+Run:
+`bazel build tensorflow/lite/micro/examples/hello_world/quantization:ptq`
+`bazel-bin/tensorflow/lite/micro/examples/hello_world/quantization:ptq  --save_dir=/tmp/model_created/`
+"""
+import math
+import os
+
+from absl import app
+from absl import flags
+from absl import logging
+import numpy as np
+import tensorflow as tf
+
+FLAGS = flags.FLAGS
+
+flags.DEFINE_integer("epochs", 500, "number of epochs to train the model.")
+flags.DEFINE_string("save_dir", "/tmp/hello_world_quant_model",
+                    "the directory to save the trained model.")
+
+
+def get_data():
+  """
+  The code will generate a set of random `x` values,calculate their sine
+  values.
+  """
+  # Generate a uniformly distributed set of random numbers in the range from
+  # 0 to 2π, which covers a complete sine wave oscillation
+  x_values = np.random.uniform(low=0, high=2 * math.pi,
+                               size=1000).astype(np.float32)
+
+  # Shuffle the values to guarantee they're not in order
+  np.random.shuffle(x_values)
+
+  # Calculate the corresponding sine values
+  y_values = np.sin(x_values).astype(np.float32)
+
+  return (x_values, y_values)
+
+
+def create_model() -> tf.keras.Model:
+  model = tf.keras.Sequential()
+
+  # First layer takes a scalar input and feeds it through 16 "neurons". The
+  # neurons decide whether to activate based on the 'relu' activation function.
+  model.add(tf.keras.layers.Dense(16, activation='relu', input_shape=(1, )))
+
+  # The new second and third layer will help the network learn more complex
+  # representations
+  model.add(tf.keras.layers.Dense(16, activation='relu'))
+
+  # Final layer is a single neuron, since we want to output a single value
+  model.add(tf.keras.layers.Dense(1))
+
+  # Compile the model using the standard 'adam' optimizer and the mean squared
+  # error or 'mse' loss function for regression.
+  model.compile(optimizer='adam', loss='mse', metrics=['mae'])
+
+  return model
+
+
+def save_tflite_model(tflite_model, save_dir, model_name):
+  """save the converted tflite model
+  Args:
+      tflite_model (binary): the converted model in serialized format.
+      save_dir (str): the save directory
+      model_name (str): model name to be saved
+  """
+  if not os.path.exists(save_dir):
+    os.makedirs(save_dir)
+  save_path = os.path.join(save_dir, model_name)
+  with open(save_path, "wb") as f:
+    f.write(tflite_model)
+  logging.info("Tflite model saved to %s", save_dir)
+
+
+def train_model(epochs, x_values, y_values):
```

Comment:
why are we re-training the model? The PTQ script should take the pre-trained model as input and only have the quantization code.

Reply:
I was initially thinking that if we just provide the ptq script without the train then the user needs to run the `train.py` script first in order to create the TF model. Thats why I was trying to make it self contained.

But it makes sense to just provide the quantization code. Now we can provide the TF model folder as a command line flag. Updated the top section of the file with relevant comments.